improve performance of GetAllNetworkInterfaces on Linux

wfurt opened the pull request:
This is somewhat related to discussion in #32611 and fixes some low-hanging issues. 
It does not change way how we get information but it avoids unnecessary processing.  

Right now when we iterate through interfaces we always do: 
1. get again routing table for v4/6 for each interface. 
- this is particularly expensive as kernel needs to walk routing table and dump it as text while processing packets.
- this can be  expensive if we ever run on system with dynamic routing
2. we read and parse /etc/resolver.conf twice for each interface
3. MTU is property of interface on Linux
- we read same value once for each protocol family. 

This change simple reads common properties once and that it used later instead if creating new copies. 

The benefit depends on number of interfaces we process. With my Ubuntu VM, I get

|Branch      |                  Method  |     Mean |     Error |    StdDev | Ratio |
|--|------------------------ |---------:|---------:|---------:|---------:|
|PR| GetAllNetworkInterfaces | 632.1 us | 12.16 us | 32.68 us | 622.5 us |
| master| GetAllNetworkInterfaces | 901.1 us | 21.86 us | 42.64 us | 883.4 us |

I also have physical machine with two interfaces and few virtual interfaces created by Docker

|Branch      |                  Method  |     Mean |     Error |    StdDev | Ratio |
|-------|------------------------ |---------:|----------:|----------:|------:|
| PR|GetAllNetworkInterfaces |  1.933 ms | 0.0112 ms | 0.0099 ms |  1.00 |
| master|GetAllNetworkInterfaces | 3.271 ms | 0.0091 ms | 0.0076 ms |  1.69 |

Note that we still throw while internally while trying to get speed for loopback and virtual interfaces. 
```
$ cat /sys/class/net/lo/speed
cat: /sys/class/net/lo/speed: Invalid argument
```

wfurt committed:
improve GetLinuxNetworkInterfaces

wfurt commented:
/azp run

stephentoub commented on src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/StringParsingHelpers.Dns.cs:
```diff
@@ -9,23 +9,21 @@ namespace System.Net.NetworkInformation
         }
 
-        internal static List<IPAddress> ParseDnsAddressesFromResolvConfFile(string filePath)
+        internal static List<IPAddress> ParseDnsAddressesFromResolvConfFile(string data)
```
Why did you change these two methods to move the File.ReadAllText to the call site?  It looks like all call sites do it now?

stephentoub replied:
Ah, I see, to avoid reading the file twice in one place where we pass the same text to two different parsers. Ok.

stephentoub commented on src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/LinuxNetworkInterface.cs:
```diff
@@ -16,15 +16,48 @@ internal class LinuxNetworkInterface : UnixNetworkInterface
+
+                try
+                {
+                    string resolverConfig = File.ReadAllText(NetworkFiles.EtcResolvConfFile);
```
In addition to the try/catch, should we also check whether the File.Exists as we do above, to avoid the exception in the failure case?  Or is that very rare?

wfurt replied:
I think that would be pretty rare. Without it system would depend only on entries in /etc/hosts. The question is if we should throw in such case or if we should return empty collection. It seems like throwing PNSP for valid configuration is not ideal IMHO.

stephentoub commented on src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/UnixIPInterfaceProperties.cs:
```diff
@@ -17,11 +17,19 @@ internal abstract class UnixIPInterfaceProperties : IPInterfaceProperties
         private readonly IPAddressCollection _dnsAddresses;
 
-        public UnixIPInterfaceProperties(UnixNetworkInterface uni)
+        public UnixIPInterfaceProperties(UnixNetworkInterface uni, bool globalConfig = false, string dnsSuffix = null, IPAddressCollection dnsAddresses = null)
```
Is this "globalConfig" needed?  Or could it just be, e.g.
```C#
_dnsSuffix = dnsSuffix ?? GetDnsSuffix();
_dnsAddresses = dnsAddresses ?? GetDnsAddresses();
```
or does that cause problems?

wfurt replied:
I did not want to try again if global attempt already fails and dnsSuffix or dnsAddresses is is null. I was focusing on Linux but perhaps best longer term fix is to update all other Unix platforms and remove  GetDns*() completely.

wfurt committed:
more improvements to GetAllNetworkInterfaces()

wfurt committed:
resolve conflicts with #41566

wfurt committed:
fix compilation

wfurt committed:
update ethtool to support older kernels

wfurt committed:
Merge branch 'interfaces' of https://github.com/wfurt/corefx into interfaces

wfurt committed:
fix osx

wfurt committed:
more platform fixes

wfurt committed:
Merge branch 'interfaces' of https://github.com/wfurt/corefx into interfaces

wfurt committed:
put PrefixLengthToSubnetMask back as it was

wfurt commented:
Thanks @stephentoub  for reviewing. I wanted to follow-up on your comments as well I had some more changes in pipeline so I decided to extend this PR instead of creating new one. 
It would be great if you can take another look. 

new nubers are:

VM

|                  Method |  Toolchain |          Mean |        
|------------------------ |---------------------------------------------------------------------------- |--------------:|
| GetAllNetworkInterfaces | master  |   1,184.41 us |  
| GetAllNetworkInterfaces |   new |     308.72 us |     

Physical Machine

|                  Method | Toolchain |          Mean |      
|------------------------ |---------------------------------------------------------------------------- |--------------:|  
| GetAllNetworkInterfaces |  master | 4,558.6 us | 
| GetAllNetworkInterfaces | new | 1,053.8 us | 


Main improvements comes from removing double pinvoke where we pinvoke native iterator to call managed callbacks. Crossing that boundary seems expensive and showed up on profiling. 

With that, I also gather various interface stats using single socket instead of opening bunch of files for each interface. I updated parsing to use Span instead of allocating array. 

Aside from various fixes for #41515, I see no reason to throw on accessing PrefixLength.
It seems like using prefix instead of netmask is cheaper and also more portable for both IPv4 and IPv6. 

I updated tests and added new one to verify netmask and PrefixLength for loopback - they should have standard value across various Linux distributions. 

fixes  #41515

wfurt committed:
normalize units properlry

wfurt committed:
fix constant

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -38,6 +44,42 @@
+    while ((mask < end) && (*mask == 0xff))
+    {
+        len += 8;
+        mask ++;
```
Nit: `mask++;`

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -38,6 +44,42 @@
 #include <net/route.h>
 #endif
 
+static inline uint8_t mask2prefix(uint8_t * mask, int length)
```
Nit: `uint8_t* mask`

stephentoub replied:
Can you add a comment explaining what this function is doing?  It's not clear to me, and in particular what the mask is and what the prefix is it's converting to (since it looks like it's returning a length).

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -38,6 +44,42 @@
+    // Get last incomplete byte
+    if (mask < end)
+    {
+        while (*mask)
```
This is probably my not understanding the purpose of the function, but... what happens if mask is much less than end?  It's ok that we may not explore the whole thing?

wfurt replied:
This is only called with NUM_BYTES_IN_IPV4_ADDRESS and NUM_BYTES_IN_IPV4_ADDRESS e.g. 4 and 16.

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -80,16 +122,11 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
                 struct sockaddr_in* mask_sain = (struct sockaddr_in*)current->ifa_netmask;
-                memcpy_s(maskInfo.AddressBytes, sizeof_member(IpAddressInfo, AddressBytes), &mask_sain->sin_addr.s_addr, sizeof(mask_sain->sin_addr.s_addr));
+                // ifa_netmask can be NULL according to documentation, probably P2P interfaces.
+                iai.PrefixLength = mask_sain != NULL ? mask2prefix((uint8_t*)&mask_sain->sin_addr.s_addr, NUM_BYTES_IN_IPV4_ADDRESS) : NUM_BYTES_IN_IPV4_ADDRESS * 8;
```
So PrefixLength uses bits as a unit?

wfurt replied:
correct. so for example 10.11.12.13 with netmask 255.255.255.0 becomes represented as 10.11.12.13/24. For IPv6 the whole netmask was dropped as impractical and only slash notation is used. New Linux tools moved towards that as well:
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 00:1c:42:20:46:36 brd ff:ff:ff:ff:ff:ff
    inet 10.211.55.17/24 brd 10.211.55.255 scope global enp0s5
       valid_lft forever preferred_lft forever
    inet6 2002:2c26:f4e4:0:21c:42ff:fe20:4636/64 scope global mngtmpaddr dynamic
       valid_lft 2591505sec preferred_lft 604305sec
    inet6 fe80::21c:42ff:fe20:4636/64 scope link
       valid_lft forever preferred_lft forever
```

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -104,10 +141,12 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
                 uint32_t scopeId = sain6->sin6_scope_id;
+
+                struct sockaddr_in6* mask_sain6 = (struct sockaddr_in6*)current->ifa_netmask;
+                iai.PrefixLength = mask_sain6 != NULL ? mask2prefix((uint8_t *)&mask_sain6->sin6_addr.s6_addr, NUM_BYTES_IN_IPV6_ADDRESS) : NUM_BYTES_IN_IPV6_ADDRESS * 8;
```
Nit: `uint8_t*`

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -158,6 +197,200 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
     return 0;
 }
 
+int32_t SystemNative_GetNetworkInterfaces(int32_t * interfaceCount, NetworkInterfaceInfo **interfaceList, int32_t * addressCount, IpAddressInfo **addressList )
```
Nit: `int32_t*`, `NetworkInterfaceInfo** `, `int32_t*`, etc. :)

stephentoub commented on src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/LinuxNetworkInterface.cs:
```diff
@@ -12,156 +14,117 @@ namespace System.Net.NetworkInformation
             {
-                lni = new LinuxNetworkInterface(name, index);
-                interfaces.Add(name, lni);
+                string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
```
GetNetworkInterfaces doesn't have `SetLastError = true` on it.

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -158,6 +197,200 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
+    ifaddrsEntry = head;
+    while (ifaddrsEntry != NULL)
+    {
+        count ++;
```
Nit: `count++;`

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -158,6 +197,200 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
+        else if (ifaddrsEntry->ifa_addr != NULL && ifaddrsEntry->ifa_addr->sa_family == AF_INET6)
+        {
+            ip6count++;
+        }
```
Nit:
```C++
if (ifaddrsEntry->ifa_addr != NULL)
{
    if (ifaddrsEntry->ifa_addr->sa_family == AF_INET)
    {
        ip4count++;
    }
    else if (ifaddrsEntry->ifa_addr->sa_family == AF_INET6)
    {
        ip6count++;
    }
}
```

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -158,6 +197,200 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
+    }
+
+    // Allocate estimated space. It can be little bit more than we need.
+    void * memoryBlock = calloc((size_t)count, sizeof(NetworkInterfaceInfo));
```
Nit: `void*`... I'll stop commenting on these.

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -158,6 +197,200 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
+    // Reset head pointers again.
+    ifaddrsEntry = head;
+    *interfaceList = nii = (NetworkInterfaceInfo*)memoryBlock;
+    *addressList = ai = (IpAddressInfo*)(nii + (count - ip4count - ip6count));
```
What is the math here doing?  It's not clear to me why indexing N from nii, where N is the number of non-IPv4/6 addresses.

wfurt replied:
I did not want to do two separate allocations. So we allocate big enough memory chunk and then we store list of IpAddressInfo behind NetworkInterfaceInfo.  I will add note to beginning where we calculate needed space.

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -158,6 +197,200 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
+        uint ifindex = if_nametoindex(ifaddrsEntry->ifa_name);
+        for (index = 0; index < (int)ifcount; index ++)
+        {
+            //if (strcmp(ni[index].Name, entry->ifa_name) == 0)
```
Delete commented out code.

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -158,6 +197,200 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
+    *addressList = ai = (IpAddressInfo*)(nii + (count - ip4count - ip6count));
+
+    while (ifaddrsEntry != NULL)
+    {
```
Can you add comments to each section here highlighting what it's doing?  I'm having trouble following the logic and will be able to review better if I know what each section of code is trying to accomplish.

stephentoub commented on src/Native/Unix/System.Native/pal_interfaceaddresses.c:
```diff
@@ -158,6 +197,200 @@ int32_t SystemNative_EnumerateInterfaceAddresses(IPv4AddressFound onIpv4Found,
+    }
+
+    // Allocate estimated space. It can be little bit more than we need.
+    void * memoryBlock = calloc((size_t)count, sizeof(NetworkInterfaceInfo));
```
Where does this eventually get freed?  If I'm following correctly, you're allocating with calloc here and then freeing with Marshal.FreeHGlobal later?  I'm not 100% sure that's valid.  @jkotas?

jkotas replied:
`calloc` is same underlying allocator as `malloc`. I do not see a problem here.

stephentoub commented on src/System.Net.NetworkInformation/tests/FunctionalTests/DnsParsingTests.cs:
```diff
@@ -29,7 +29,7 @@ public void DnsAddressesParsing(string file)
             FileUtil.NormalizeLineEndings(file, fileName);
 
-            var dnsAddresses = StringParsingHelpers.ParseDnsAddressesFromResolvConfFile(fileName);
+            var dnsAddresses = StringParsingHelpers.ParseDnsAddressesFromResolvConfFile(File.ReadAllText(fileName));
```
Nit: var => ?

jkotas commented on src/Common/src/Interop/Unix/System.Native/Interop.EnumerateInterfaceAddresses.cs:
```diff
@@ -24,10 +24,26 @@ public unsafe struct IpAddressInfo
+            public byte NumAddressBytes;
+            public fixed byte AddressBytes[8];
+            public byte SupportsMulticast;
+             private fixed byte __padding[3];
```
```suggestion
            private fixed byte __padding[3];
```

wfurt committed:
feedback from review

wfurt committed:
fix log initialization
